content-src/components/TopSites/SearchShortcutsForm.jsx

```diff
@@ -0,0 +1,143 @@
+import {actionCreators as ac, actionTypes as at} from "common/Actions.jsm";
+import {FormattedMessage} from "react-intl";
+import React from "react";
+
+class SelectableSearchShortcut extends React.PureComponent {
+  render() {
+    const {shortcut, selected} = this.props;
+    const imageStyle = {backgroundImage: `url("${shortcut.icon}")`};
+    return (
+      <div className="top-site-outer">
+        <input type="checkbox" id={shortcut.keyword} name={shortcut.keyword} checked={selected} onChange={this.props.onChange} />
+        <label htmlFor={shortcut.keyword}>
+          <div className="top-site-inner">
+            <span>
+              <div className="tile">
+                <div className="top-site-icon rich-icon" style={imageStyle} data-fallback="@" />
+                <div className="top-site-icon search-topsite" />
+              </div>
+              <div className="title">
+                <span dir="auto">{shortcut.keyword}</span>
+              </div>
+            </span>
+          </div>
+        </label>
+      </div>
+    );
+  }
+}
+
+export class SearchShortcutsForm extends React.PureComponent {
+  constructor(props) {
+    super(props);
+    this.handleChange = this.handleChange.bind(this);
+    this.onCancelButtonClick = this.onCancelButtonClick.bind(this);
+    this.onAddButtonClick = this.onAddButtonClick.bind(this);
+
+    // clone the shortcuts and add them to the state so we can add isSelected property
+    const shortcuts = [];
+    const {rows, searchShortcuts} = props.TopSites;
+    searchShortcuts.forEach(shortcut => {
+      shortcuts.push(Object.assign({}, shortcut, {isSelected: !!rows.find(({isPinned, searchTopSite, label}) => isPinned && searchTopSite && label === shortcut.keyword)}));
+    });
+    this.state = {shortcuts};
+  }
+
+  handleChange(event) {
+    const {target} = event;
+    const {name, checked} = target;
+    this.setState(prevState => {
+      const shortcuts = prevState.shortcuts.slice();
+      let shortcut = shortcuts.find(({keyword}) => keyword === name);
+      shortcut.isSelected = checked;
+      return {shortcuts};
+    });
+  }
+
+  onCancelButtonClick(ev) {
+    ev.preventDefault();
+    this.props.onClose();
+  }
+
+  onAddButtonClick(ev) {
+    ev.preventDefault();
+
+    // Check if there were any changes and act accordingly
+    const {rows} = this.props.TopSites;
+    const pinQueue = [];
+    const unpinQueue = [];
+    this.state.shortcuts.forEach(shortcut => {
+      const alreadyPinned = rows.find(({isPinned, searchTopSite, label}) => isPinned && searchTopSite && label === shortcut.keyword);
+      if (shortcut.isSelected && !alreadyPinned) {
+        pinQueue.push(this._searchTopSite(shortcut));
+      } else if (!shortcut.isSelected && alreadyPinned) {
+        unpinQueue.push({url: alreadyPinned.url});
+      }
```

Comment:
All this pinning and unpinning logic might want to be in a feed instead. But.... good enough for now?

Reply:
Can we not just insert into the first position? Is the dialog supposed to show already pinned stuff?

Reply:
I assumed it would always show all the options available?